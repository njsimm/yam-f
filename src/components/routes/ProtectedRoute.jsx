import React, { useContext, useEffect, useState } from "react";
import { Navigate, Outlet, useParams } from "react-router-dom";
import UserContext from "../../utils/UserContext";
import YamAPI from "../../utils/YamApi";
import CircularProgress from "@mui/material/CircularProgress";
import Box from "@mui/material/Box";

export default function ProtectedRoute() {
  const { currentUser } = useContext(UserContext);
  const { userId, productId, businessId, saleId, businessSaleId } = useParams();
  const [loading, setLoading] = useState(true);
  const [authorized, setAuthorized] = useState(true);

  useEffect(() => {
    const checkAuthorization = async () => {
      setLoading(true);
      setAuthorized(true);

      if (!currentUser) {
        setAuthorized(false);
        setLoading(false);
        return;
      }

      if (userId && parseInt(userId, 10) !== currentUser.id) {
        setAuthorized(false);
        setLoading(false);
        return;
      }

      try {
        if (productId) {
          const product = await YamAPI.getProductById(
            currentUser.id,
            productId
          );
          if (product.userId !== currentUser.id) {
            setAuthorized(false);
          }
        } else if (businessId) {
          const business = await YamAPI.getBusinessById(
            currentUser.id,
            businessId
          );
          if (business.userId !== currentUser.id) {
            setAuthorized(false);
          }
        } else if (saleId) {
          const sale = await YamAPI.getSaleById(productId, saleId);
          if (sale.userId !== currentUser.id) {
            setAuthorized(false);
          }
        } else if (businessSaleId) {
          const businessSale = await YamAPI.getBusinessSaleById(
            businessId,
            businessSaleId
          );
          if (businessSale.userId !== currentUser.id) {
            setAuthorized(false);
          }
        }
      } catch (error) {
        setAuthorized(false);
      }

      setLoading(false);
    };

    checkAuthorization();
  }, [currentUser, userId, productId, businessId, saleId, businessSaleId]);

  if (loading) {
    return (
      <Box
        sx={{
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
          height: "100vh",
        }}
      >
        <CircularProgress />
      </Box>
    );
  }

  if (!authorized) {
    return <Navigate to="/not-authorized" />;
  }

  return <Outlet />;
}
