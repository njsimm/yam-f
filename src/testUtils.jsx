import React from "react";
import { render } from "@testing-library/react";
import UserContext from "./utils/UserContext";
import SidebarContext from "./utils/SidebarContext";
import { vi } from "vitest"; // Import vi from vitest

// Mock user data
export const mockUser = {
  id: 1,
  email: "demo@example.com",
  username: "demoUser",
  password: "password123",
  first_name: "Demo",
  last_name: "User",
  is_admin: false,
  date_created: "2024-06-24T12:00:00.000Z",
};

// Mock sidebar context data
export const mockSidebarContext = {
  isSidebarOpen: true,
  toggleSidebar: vi.fn(),
};

// UserProvider component to wrap components with UserContext
const UserProvider = ({ children, currentUser = mockUser }) => {
  return (
    <UserContext.Provider value={{ currentUser }}>
      {children}
    </UserContext.Provider>
  );
};

// SidebarProvider component to wrap components with SidebarContext
const SidebarProvider = ({ children, sidebarContext = mockSidebarContext }) => {
  return (
    <SidebarContext.Provider value={sidebarContext}>
      {children}
    </SidebarContext.Provider>
  );
};

const AllProviders = ({
  children,
  currentUser = mockUser,
  sidebarContext = mockSidebarContext,
}) => {
  return (
    <UserProvider currentUser={currentUser}>
      <SidebarProvider sidebarContext={sidebarContext}>
        {children}
      </SidebarProvider>
    </UserProvider>
  );
};

export * from "@testing-library/react";
export { UserProvider, SidebarProvider, AllProviders };
