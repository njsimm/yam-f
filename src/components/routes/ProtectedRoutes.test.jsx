import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import ProtectedRoute from "./ProtectedRoute";
import { AllProviders, mockUser } from "../../testUtils";
import { expect, test } from "vitest";

test("renders protected Dashboard component for authenticated user", () => {
  const { asFragment } = render(
    <MemoryRouter initialEntries={["/users/dashboard"]}>
      <AllProviders currentUser={mockUser}>
        <Routes>
          <Route element={<ProtectedRoute />}>
            <Route path="/users/dashboard" element={<div>Dashboard</div>} />
          </Route>
        </Routes>
      </AllProviders>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

test("redirects to Not Authorized for unauthenticated user", () => {
  const { asFragment } = render(
    <MemoryRouter initialEntries={["/users/dashboard"]}>
      <AllProviders currentUser={null}>
        <Routes>
          <Route element={<ProtectedRoute />}>
            <Route path="/users/dashboard" element={<div>Dashboard</div>} />
          </Route>
          <Route path="/not-authorized" element={<div>Not Authorized</div>} />
        </Routes>
      </AllProviders>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});
