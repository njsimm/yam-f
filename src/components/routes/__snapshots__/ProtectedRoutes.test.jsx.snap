// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`redirects to Not Authorized for unauthenticated user 1`] = `
<DocumentFragment>
  <div>
    Not Authorized
  </div>
</DocumentFragment>
`;

exports[`renders protected Dashboard component for authenticated user 1`] = `
<DocumentFragment>
  <div>
    Dashboard
  </div>
</DocumentFragment>
`;
